### DISCUSSION TITLE
using PyVista from pure-python web framework

### DISCUSSION THREAD

**Discussion by gsal** [pyvista/pyvista#4748] [Q&A] [answered]
Lately, development is taking a turn towards the browser, web-based apps, the cloud, whatever you call it.

I am not a full stack developer but, thankfully, quite a few pure-python web-development frameworks have popped up: nicegui, solara, reflex, etc.

I looked at Trame but seems a bit convoluted; sure, the vtk widget is well documented, but when it comes to the other widgets typical of a dashboard application, they seem to rely in Vuetify and one needs to look the docs for javascript and convert to python...not cool.

Could I use PyVista from within one of these web frameworks? I do see mention of a trame backend. But, in general, web-based use of PyVista is not very clear in the documentation or, somehow, jupyter-notebook oriented.

For frontend widgets and the like, I find something like reflex and nicegui much easier to use than trame; but I also want to take advantage of VTK, somehow. 

And so, would it be too much to ask for a PyVista example for the browser that shows the same minimal trame/examples/06_vtk/01_SimpleCone ? 

Thanks.

**Comment by tkoyama010** [comment]
Earlier PyVista had flask examples, but these have been removed.
https://github.com/pyvista/pyvista/pull/4299
The flask examples were deemed to lag behind the Trame examples. If you can show a good use for PyVista in a web framework we can discuss it.

**Comment by gsal** [comment]
Wait, what? Are you asking me what I asked "you"?  I asked first!

Or

I don't know what you mean. 

Otherwise, I am asking for an example python program that shows the use of pyvista in a non-jupyter, web-based context.

**Reply by tkoyama010** [reply]
Sorry, I mistakenly thought you were asking about adding your own web app example to PyVista. How about using the Panel from the Holoviz project, as shown in this example?
https://panel.holoviz.org/gallery/vtk_interactive.html
As it uses vtkjs, it does not have full vtk support like Trame does, but it can do something similar.

Edit: There are currently no plans to add any web frameworks other than Trame to the PyVista project document itself.

**Comment by gsal** [comment]
Ha, ha, no worries.

Here is my confusion...the way I see it, pyVista already:
- uses VTK
- uses Trame and has a backend
- renders in a browser, albeit inside a jupyter notebook

So, how come it cannot render in a browser, outside a jupyter notebook?

**Reply by tkoyama010** [reply, marked as answer]
The following thread will answer your question.
https://discourse.holoviz.org/t/trame-support-in-panel-to-fully-support-vtk/5275/1
The conclusion is that PyVista relies on the help of ipywidgets to plot on Jupyter. Therefore, in order to display PyVista in the browser, a framework is needed to display the output of ipywidgets in the browser, such as Panel from the Holoviz project.
Edit: Here is the code and output.
```python
import panel as pn
import pyvista as pv
from IPython.display import IFrame
from pyvista import examples


def handler(viewer, src, **kwargs):
    return IFrame(src, "75%", "500px")


# Load the statue mesh
mesh = examples.download_nefertiti()
mesh.rotate_x(-90.0, inplace=True)  # rotate to orient with the skybox

# Download skybox
cubemap = examples.download_sky_box_cube_map()

p = pv.Plotter(notebook=True)
p.add_actor(cubemap.to_skybox())
p.set_environment_texture(cubemap)  # For reflecting the environment off the mesh
p.add_mesh(mesh, color="linen", pbr=True, metallic=0.8, roughness=0.1, diffuse=1)

# Define a nice camera perspective
cpos = [(-313.40, 66.09, 1000.61), (0.0, 0.0, 0.0), (0.018, 0.99, -0.06)]

iframe = p.show(
    jupyter_backend="trame",
    jupyter_kwargs=dict(handler=handler),
    return_viewer=True,
    cpos=cpos,
)

pn.extension()
pn.panel(iframe, width=2000).show()
```
[Screencast from 2023-05-03 16-09-10.webm](https://github.com/pyvista/pyvista/assets/7513610/2684ef46-c531-48f9-aad3-15b5211fc4f7)

**Reply by gsal** [reply]
Thank very much.

**Reply by Jupihawx** [reply]
Great thanks for the code.
I was just wondering, is there a way to host the result on the web, instead of localhost? I tried with ngrok for a quick-and-dirty method but to no avail.
I actually managed with localtunnel, but since the app calls another port for the backend I believe, i get a localhost connection refused (which make sense since only the frontend is served)
